File: main/src/main/java/com/sedmelluq/discord/lavaplayer/track/BaseAudioTrack.java
package com.sedmelluq.discord.lavaplayer.track;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManager;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioFrame;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioTrackExecutor;
import com.sedmelluq.discord.lavaplayer.track.playback.MutableAudioFrame;
import com.sedmelluq.discord.lavaplayer.track.playback.PrimordialAudioTrackExecutor;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Abstract base for all audio tracks with an executor
 */
public abstract class BaseAudioTrack implements InternalAudioTrack {
  private final PrimordialAudioTrackExecutor initialExecutor;
  private final AtomicBoolean executorAssigned;
  private volatile AudioTrackExecutor activeExecutor;
  protected final AudioTrackInfo trackInfo;
  protected final AtomicLong accurateDuration;
  private volatile Object userData;

  /**
   * @param trackInfo Track info
   */
  public BaseAudioTrack(AudioTrackInfo trackInfo) {
    this.initialExecutor = new PrimordialAudioTrackExecutor(trackInfo);
    this.executorAssigned = new AtomicBoolean();
    this.activeExecutor = null;
    this.trackInfo = trackInfo;
    this.accurateDuration = new AtomicLong();
  }

  @Override
  public void assignExecutor(AudioTrackExecutor executor, boolean applyPrimordialState) {
    if (executorAssigned.compareAndSet(false, true)) {
      if (applyPrimordialState) {
        initialExecutor.applyStateToExecutor(executor);
      }
      activeExecutor = executor;
    } else {
      throw new IllegalStateException("Cannot play the same instance of a track twice, use track.makeClone().");
    }
  }

  @Override
  public AudioTrackExecutor getActiveExecutor() {
    AudioTrackExecutor executor = activeExecutor;
    return executor != null ? executor : initialExecutor;
  }

  @Override
  public void stop() {
    getActiveExecutor().stop();
  }

  @Override
  public AudioTrackState getState() {
    return getActiveExecutor().getState();
  }

  @Override
  public String getIdentifier() {
    return trackInfo.identifier;
  }

  @Override
  public boolean isSeekable() {
    return !trackInfo.isStream;
  }

  @Override
  public long getPosition() {
    return getActiveExecutor().getPosition();
  }

  @Override
  public void setPosition(long position) {
    getActiveExecutor().setPosition(position);
  }

  @Override
  public void setMarker(TrackMarker marker) {
    getActiveExecutor().setMarker(marker);
  }

  @Override
  public AudioFrame provide() {
    return getActiveExecutor().provide();
  }

  @Override
  public AudioFrame provide(long timeout, TimeUnit unit) throws TimeoutException, InterruptedException {
    return getActiveExecutor().provide(timeout, unit);
  }

  @Override
  public boolean provide(MutableAudioFrame targetFrame) {
    return getActiveExecutor().provide(targetFrame);
  }

  @Override
  public boolean provide(MutableAudioFrame targetFrame, long timeout, TimeUnit unit)
      throws TimeoutException, InterruptedException {

    return getActiveExecutor().provide(targetFrame, timeout, unit);
  }

  @Override
  public AudioTrackInfo getInfo() {
    return trackInfo;
  }

  @Override
  public long getDuration() {
    long accurate = accurateDuration.get();

    if (accurate == 0) {
      return trackInfo.length;
    } else {
      return accurate;
    }
  }

  @Override
  public AudioTrack makeClone() {
    throw new UnsupportedOperationException();
  }

  @Override
  public AudioSourceManager getSourceManager() {
    return null;
  }

  @Override
  public AudioTrackExecutor createLocalExecutor(AudioPlayerManager playerManager) {
    return null;
  }

  @Override
  public void setUserData(Object userData) {
    this.userData = userData;
  }

  @Override
  public Object getUserData() {
    return userData;
  }

  @Override
  @SuppressWarnings("unchecked")
  public <T> T getUserData(Class<T> klass) {
    Object data = userData;

    if (data != null && klass.isAssignableFrom(data.getClass())) {
      return (T) data;
    } else {
      return null;
    }
  }
}

File: main/src/main/java/lavaplayer/track/BaseAudioTrack.kt
package lavaplayer.track

import lavaplayer.manager.AudioPlayerManager
import lavaplayer.source.ItemSourceManager
import lavaplayer.track.playback.AudioFrame
import lavaplayer.track.playback.AudioTrackExecutor
import lavaplayer.track.playback.MutableAudioFrame
import lavaplayer.track.playback.PrimordialAudioTrackExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * Abstract base for all audio tracks with an executor
 *
 * @param info Track info
 */
abstract class BaseAudioTrack(final override val info: AudioTrackInfo) : InternalAudioTrack {
    @JvmField
    protected val accurateDuration = AtomicLong()

    private val initialExecutor = PrimordialAudioTrackExecutor(info)
    private val executorAssigned = AtomicBoolean()
    @Volatile
    private var _activeExecutor: AudioTrackExecutor? = null
    @Volatile
    override var userData: Any? = null

    override val activeExecutor: AudioTrackExecutor
        get() = _activeExecutor ?: initialExecutor

    override val sourceManager: ItemSourceManager?
        get() = null

    override val state: AudioTrackState
        get() = activeExecutor.state

    override val identifier: String
        get() = info.identifier

    override val isSeekable: Boolean
        get() = !info.isStream

    override var position: Long
        get() = activeExecutor.position
        set(position) {
            activeExecutor.position = position
        }

    override val duration: Long
        get() {
            val accurate = accurateDuration.get()
            return if (accurate == 0L) info.length else accurate
        }

    override fun assignExecutor(executor: AudioTrackExecutor?, applyPrimordialState: Boolean) {
        _activeExecutor = if (executorAssigned.compareAndSet(false, true)) {
            if (applyPrimordialState) {
                initialExecutor.applyStateToExecutor(executor)
            }

            executor
        } else {
            throw IllegalStateException("Cannot play the same instance of a track twice, use track.makeClone().")
        }
    }

    override fun stop() {
        activeExecutor.stop()
    }

    override fun setMarker(marker: TrackMarker?) {
        activeExecutor.setMarker(marker)
    }

    override fun provide(): AudioFrame? {
        return activeExecutor.provide()
    }

    @Throws(TimeoutException::class, InterruptedException::class)
    override fun provide(timeout: Long, unit: TimeUnit): AudioFrame? {
        return activeExecutor.provide(timeout, unit)
    }

    override fun provide(targetFrame: MutableAudioFrame): Boolean {
        return activeExecutor.provide(targetFrame)
    }

    @Throws(TimeoutException::class, InterruptedException::class)
    override fun provide(targetFrame: MutableAudioFrame, timeout: Long, unit: TimeUnit): Boolean {
        return activeExecutor.provide(targetFrame, timeout, unit)
    }

    override fun makeClone(): AudioTrack? {
        val track = makeShallowClone()
        track.userData = userData
        return track
    }

    override fun createLocalExecutor(playerManager: AudioPlayerManager?): AudioTrackExecutor? {
        return null
    }

    override fun <T> getUserData(klass: Class<T>?): T? {
        val data = userData
        return if (data != null && klass!!.isAssignableFrom(data.javaClass)) data as? T else null
    }

    protected open fun makeShallowClone(): AudioTrack {
        throw UnsupportedOperationException()
    }
}
